oqsprov/oqs_kem.c

```diff
@@ -89,51 +95,289 @@ static int oqs_kem_decaps_init(void *vpkemctx, void *vkem, const OSSL_PARAM para
+
+    pkeylen = EVP_PKEY_get1_encoded_public_key(pkey, &ctkex_encoded);
+    ON_ERR_SET_GOTO(pkeylen <= 0 || !ctkex_encoded || pkeylen != pubkey_kexlen, ret, -1, err);
+
+    memcpy(ct, ctkex_encoded, pkeylen);
+
+    err:
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_CTX_free(kgctx);
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_free(peerpk);
+    OPENSSL_free(ctkex_encoded);
+    return ret;
+}
+
+static int oqs_evp_kem_decaps_keyslot(void *vpkemctx, unsigned char *secret, size_t *secretlen,
+                                      const unsigned char *ct, size_t ctlen, int keyslot)
+{
+    OQS_KEM_PRINTF("OQS KEM provider called: oqs_hyb_kem_decaps\n");
+
+    int ret = OQS_SUCCESS, ret2 = 0;
+    const PROV_OQSKEM_CTX *pkemctx = (PROV_OQSKEM_CTX *)vpkemctx;
+    const OQSX_EVP_CTX *evp_ctx = pkemctx->kem->oqsx_provider_ctx.oqsx_evp_ctx;
+
+    size_t pubkey_kexlen = evp_ctx->kex_info->kex_length_public_key;
+    size_t kexDeriveLen = evp_ctx->kex_info->kex_length_secret;
+    unsigned char *privkey_kex = pkemctx->kem->privkey_ptr[keyslot];
+    size_t privkey_kexlen = evp_ctx->kex_info->kex_length_private_key;
+
+    // Free at err:
+    EVP_PKEY_CTX *ctx;
+    EVP_PKEY *pkey = NULL, *peerpkey = NULL;
+
+    *secretlen = kexDeriveLen;
+    if (secret == NULL) return 1;
+
+    if (evp_ctx->kex_info->raw_key_support) {
+        pkey = EVP_PKEY_new_raw_private_key(evp_ctx->kex_info->nid_kex, NULL, privkey_kex, privkey_kexlen);
+        ON_ERR_SET_GOTO(!pkey, ret, -10, err);
+    } else {
+        pkey = d2i_AutoPrivateKey(&pkey, (const unsigned char **)&privkey_kex, privkey_kexlen);
+        ON_ERR_SET_GOTO(!pkey, ret, -2, err);
+    }
+
+    peerpkey = EVP_PKEY_new();
+    ON_ERR_SET_GOTO(!peerpkey, ret, -3, err);
+
+    ret2 = EVP_PKEY_copy_parameters(peerpkey, evp_ctx->kexParam);
+    ON_ERR_SET_GOTO(ret2 <= 0, ret, -4, err);
+
+    ret2 = EVP_PKEY_set1_encoded_public_key(peerpkey, ct, pubkey_kexlen);
+    ON_ERR_SET_GOTO(ret2 <= 0 || !peerpkey, ret, -5, err);
+
+    ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    ON_ERR_SET_GOTO(!ctx, ret, -6, err);
+
+    ret = EVP_PKEY_derive_init(ctx);
+    ON_ERR_SET_GOTO(ret <= 0, ret, -7, err);
+    ret = EVP_PKEY_derive_set_peer(ctx, peerpkey);
+    ON_ERR_SET_GOTO(ret <= 0, ret, -8, err);
+
+    ret = EVP_PKEY_derive(ctx, secret, &kexDeriveLen);
+    ON_ERR_SET_GOTO(ret <= 0, ret, -9, err);
+
+    err:
+    EVP_PKEY_free(peerpkey);
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(ctx);
+    return ret;
+}
+
+/// Hybrid KEM functions
+
+static int oqs_hyb_kem_encaps(void *vpkemctx, unsigned char *ct, size_t *ctlen,
+                              unsigned char *secret, size_t *secretlen)
+{
+    int ret = OQS_SUCCESS;
+    const PROV_OQSKEM_CTX *pkemctx = (PROV_OQSKEM_CTX *)vpkemctx;
+    size_t secretLen0 = 0, secretLen1 = 0;
+    size_t ctLen0 = 0, ctLen1 = 0;
+    unsigned char *ct0, *ct1, *secret0, *secret1;
+
+    ret = oqs_evp_kem_encaps_keyslot(vpkemctx, NULL, &ctLen0, NULL, &secretLen0, 0);
```

Comment:
I very much like the keyslot idea. Would it be sensible to drive it further by setting up an array _defining_ the hybrid scheme? Something like `[Alg_Name0, Alg_Name1, Alg_Name2, ...]` (or IDs) and call in loop(s) `oqs_qs_kem_encaps_keyslot` or `oqs_evp_kem_encaps_keyslot` (or ...decaps... below) as appropriate? Just an idea, surely not essential for this PR.

Reply:
Yes this would be useful, something more generic like this was also my initial idea when adding the keyslots. The main obstacle right now is that secret and ciphertext lengths are attached to source-specific structs (i.e. OQS_KEM/OQS_SIG vs the custom structs I've added for EVP). If this can be unified (e.g.  by queries to the new provider API), your suggested approach should be doable.